Add axes labels to response plots

valentin-krasontovitsch opened the pull request:
**Issue**
Resolves #247

**Approach**
Hard code axes labels "Data" and "Value" for all response graphs.


## Pre review checklist

- [x] Added appropriate labels

**not sure if I should also label this with `milestone` and / or `improvement`**

Also no idea what kind of labels could be added to the axes of parameter plots. I welcome any hints and suggestions!

ertomatic commented:
Can one of the admins verify this patch?

valentin-krasontovitsch commented:
playing a bit, i observe the following behaviour:

a) adding one or more parameters, clearing all parameters removes all graphs works fine
b) adding zero or more parameters, one response, and one or more parameters, clearing all parameters clears only the first parameter graph
c) in scenario b), clearing the response after clearing all parameters removes all remaining graphs

valentin-krasontovitsch commented:
ok found it, indeed just the first plot that didn't match the current selection of params / responses was deleted, so deseleting a single entity would always work, while clearing multiple selected items would lead to the observed behaviour.

valentin-krasontovitsch commented:
and indeed, after removing parameters, when one removes all responses, the current selection, which constitutes the input for the plot grid creation, was and gets updated correctly to an empty list, and the function `create_grid` returns an empty list as well - all observed behaviour can be explained.

DanSava commented on webviz_ert/assets/ert-style.json:
```diff
@@ -131,6 +131,18 @@
+        "layout-date-value-axis-labels": {
+            "xaxis": {
+              "title": {
+                "text": "Date"
```
The x_axis is not always a datetime, in the poly_example the index is just a numerical index. Is it possible to change the label based on the index type? `Date` for datetime and maybe just `Index` for numerical values. 

Or maybe somebody has a better idea than `Index` @mortalisk @frode-aarstad @oysteoh

valentin-krasontovitsch replied:
I'll look into it and see if I can access the necessary info! thanks for pointing out the example

valentin-krasontovitsch replied:
so looking at the code, I can see where I can access the info - the `response` of type `webviz.models.Response` has a property `data` that contains a panda dataframe. the `index` of the dataframe has a type - in the case of poly_example, the type is `pandas.core.indexes.numeric.Int64Index`, but for snake_oil_field, the type is just a generic `pandas.core.indexes.base.Index`.

a possible heuristic at this point would be to set a label "Index" when the type is Int64Index, and "Date" otherwise.

one thing that makes me hesitate with implementing this is that the properties i need are not quite readily available / accessible where I want to use them. maybe i don't quite see how i should use the code, maybe it needs a bit of restructuring - in any case, i would love some input on this, maybe we could have a chat about this tomorrow?

valentin-krasontovitsch replied:
after a live discussion, Dan pointed out an elegant solution to my problem.

now all that's left is to add some test(s) that cover this - it seems like i will need to add some test data in `tests/data/snake_oil_data.py` ... too late to try and wrap my head around that now

valentin-krasontovitsch replied:
resolving this as the label is dynamic now

oysteoh commented:
Test this please!

DanSava commented on tests/data/snake_oil_data.py:
```diff
@@ -12,6 +12,24 @@ def content(self):
         return self._content
 
 
+ensemble_for_axis_labels_tests = "default3"
+axis_labels_test_index_response = "FGPT"
+axis_labels_test_date_response = "FOPR"
```
I would just use the strings and not add these variables. Adding them dose not improve code readability.

For example, now you write:

```
all_ensemble_names = [
    "default",
    "default_smoother_update",
    ensemble_for_axis_labels_tests,
    "nr_42",
]
```
and the above list contains 3 strings and the variable `ensemble_for_axis_labels_tests` this does not seem very consistent.

valentin-krasontovitsch replied:
Hm, yes, I agree - but I wasn't going for readability yet. This is an attempt to move towards data driven tests. 
I understand it makes for incohesive style in the code right now, and I'm hoping that this can be resolved with a refactoring PR as hinted at above.
also, it's mostly DRYness within the same file I'm thinking of with lines 15-17.
furthermore - this (the variable `all_ensemble_names`) actually serves to reduce brittleness. one unrelated test broke, that chose certain ensembles, and expected to just find another ensemble left, all hard coded.

please let me know if i was successful with defending this position.

valentin-krasontovitsch replied:
right, I too think hard coding them in the test is OK - I just don't think there's a need to do it many times, once should be enough.

I don't know, maybe this is a bit philosophical, I see the point of loading the ensemble names like you suggest. When hard-coding the list, I think I had in mind what you mention with having it explicit in the code, so that you notice when something breaks.

And I like hyper-explicit names : D it is strange to use that name in another test, and I would solve that problem by doing something like

```
ensemble_for_that_other_kind_of_test = ensemble_for_axis_labels_test
```

(which gives you a clear dependency, which one might wonder is that something you want, ...)

OK so let me summarize a bit:

- I don't want to use the literal string more than once in the test file -> put it in variable
- I don't mind much about the name of the variable, so I can change that, if desired
- I still think that it's reasonable to use that list of ensembles in the example where I used it, because if you leave it as it was, by changing the test data like I did, you break that test in a "formal" way, you're not breaking anything meaningful or functional.
- How we define the list of ensembles, I don't really mind I suppose, we can do it dynamically via that API call, or have it hard-coded in a list...

what do you think? can i leave the variables? should i change the names and / or the definition of the list? or should I scrape it (the variables) all, and leave it as literal strings everywhere, and the literal list of ensembles in that one test?

valentin-krasontovitsch replied:
not sure if you wanted to add anything here, @DanSava, so I'm gonna leave it open for now

DanSava replied:
> what do you think? can i leave the variables? should i change the names and / or the definition of the list? or should I scrape it (the variables) all, and leave it as literal strings everywhere, and the literal list of ensembles in that one test?

I think you should do what makes you happy what makes the code logical and easy to understand what makes the code readable and achieve this by removing logic from the codebase not adding it, in this order. 

I guess what I am saying is do what makes you happy :smiley: 

 And if you like hyper-explicit names I would suggest using something similar in each test:

```
ensemble_for_axis_labels_test = ensemble_general_constant_varaible_name_defined_somewhere_in_the_conftest_file
```

valentin-krasontovitsch replied:
ok thank you : )

valentin-krasontovitsch replied:
gonna resolve this

oysteoh commented on tests/views/helpers.py:
```diff
@@ -0,0 +1,79 @@
+    ensemble_selector.click()
+    dash_duo.wait_for_contains_text(
+        "#" + plugin.uuid("selected-ensemble-dropdown"),
+        wanted_ensemble_name,
+        timeout=4,
+    )
+
+
+def select_response(dash_duo, plugin, response_name):
+    _select_param_or_response(
+        dash_duo,
+        plugin,
+        f'#{plugin.uuid("parameter-selector-multi-resp")}',
+        response_name,
+    )
+
+
+def _select_param_or_response(dash_duo, plugin, containerSelector, option_name):
+    selectors = dash_duo.find_elements(f"{containerSelector} option")
+    wanted_option_selector = list(
+        filter(lambda rs_sel: rs_sel.text == option_name, selectors)
+    )[0]
+    wanted_option_selector.click()
+    dash_duo.wait_for_element(f"#{plugin.uuid(option_name)}")
+
+
+def select_parameter(dash_duo, plugin, parameter_name):
+    _select_param_or_response(
+        dash_duo,
+        plugin,
+        f'#{plugin.uuid("parameter-selector-multi-param")}',
+        parameter_name,
+    )
+
+
+def wait_a_bit(dash_duo, time_seconds=0.1):
+    try:
+        dash_duo.wait_for_element(".foo-baarrrrr", timeout=time_seconds)
+    except TimeoutException:
+        pass
```
As an alternative to a separate PR it would also be ok to me at least having separate commits in the same PR. One solving the actual issue and then the refactoring afterwards. And yes - i fully agree to your comment keeping the changes needed to solve the PR to a minimum :+1: :+1:

valentin-krasontovitsch replied:
@DanSava - thanks for the pointer with the `helper` file naming, I'm very much not versed in the coding style and best practice when it comes to python testing. fixtures, I'll have to read a bit more about, but i have a feeling it's not super relevant in this case?
of course I can move these functions to the conftest file. can a conftest file reside at different scopes (in different directories), or can there only be one?

also - if you have any links to best practice for testing with fixtures and conftest files and whatnot flying around / in your history, would you mind sharing some?

You're right of course, I see now that I went a bit overboard with applying these helper functions outside of the code I added - mea culpa. I'm gonna do these changes as a clear-cut separate commit, as @oysteoh suggests (thanks for that!)

valentin-krasontovitsch replied:
after reading a bit more of the code in `tests/conftest.py`, I see that the helper functions i wrote are somewhat similar to `select_first` and `select_by_name` - should I try to refactor my helpers so they use those? I guess that would reduce the complexity and make the code overall DRYer

valentin-krasontovitsch replied:
ah `select_by_name` is a new function : )

DanSava replied:
> of course I can move these functions to the conftest file. can a conftest file reside at different scopes (in different directories), or can there only be one?


Yes a conftest file can reside at different scopes, also there can be multiple of these files more info here : https://docs.pytest.org/en/6.2.x/fixture.html#conftest-py-sharing-fixtures-across-multiple-files
But for a project as complex as webviz-ert one conftest file is more than enough. 


> after reading a bit more of the code in tests/conftest.py, I see that the helper functions i wrote are somewhat similar to select_first and select_by_name - should I try to refactor my helpers so they use those? I guess that would reduce the complexity and make the code overall DRYer


Yes, I think it is a good idea to try and use as much as possible the already implemented used and tested logic that is present before jumping and adding a bunch of new logic that might just be similar.

valentin-krasontovitsch replied:
hehe, I realized I was already using `find_first`, and we basically did the exact same thing in the other helper function ^^ that was an easy refactor!

valentin-krasontovitsch commented on tests/data/snake_oil_data.py:
```diff
@@ -12,6 +12,24 @@ def content(self):
         return self._content
 
 
+ensemble_for_axis_labels_tests = "default3"
+axis_labels_test_index_response = "FGPT"
+axis_labels_test_date_response = "FOPR"
```
closing in favor of (maybe duplicate?) other discussion further down

DanSava commented on tests/data/snake_oil_data.py:
```diff
@@ -54,10 +66,10 @@ def content(self):
                         {
                             "children": [],
-                            "name": "default3",
+                            "name": ensemble_for_axis_labels_tests,
```
Is it possible to just use the hard-coded strings for now? Or just use a different variable name, something like `ensemble_name_3`

valentin-krasontovitsch replied:
yeah sure, I can do `ensemble_name_3` : )

valentin-krasontovitsch replied:
changed the naming to be a bit less specific.

valentin-krasontovitsch commented:
OK I've rebased, everything should look neat now, I discovered that I had not implemented the dynamic x axis labelling for the response correlation viewer and fixed that, and covered it with a test, and now it should be OK.

would somebody be so kind and have another look please? I probably won't be able to react until monday, just fyi

DanSava commented on tests/data/snake_oil_data.py:
```diff
@@ -145,9 +163,14 @@ def content(self):
-        "FOPR": {
-            "name": "FOPR",
-            "id": "FOPR",
+        f"{INDEX_RESPONSE}": {
```
I think : `f"{INDEX_RESPONSE}": {...}` is equivalent with `INDEX_RESPONSE: {...}`

valentin-krasontovitsch replied:
yes, of course, thank you : )
fixed!

oysteoh commented:
Jenkins: Test this please!

valentin-krasontovitsch commented:
OK it should be ready for merge now - all comments reacted on, nicely cut commits (work + refactoring), and fresh rebase onto  current main

it says there are changes requested even though I've tried to resolve all comments - could you, @DanSava, kindly have a quick look please? I guess it just needs an approve button click...

not sure what the jenkins thing is though - gonna try to rerun

valentin-krasontovitsch commented:
Jenkins: Test this please!

valentin-krasontovitsch commented:
yeahy apparently jenkins is happy now : )

DanSava commented on tests/conftest.py:
```diff
@@ -100,3 +101,68 @@ def select_by_name(dash_duo, selector, name):
+    return wanted_ensemble_name
+
+
+def _select_param_or_response(dash_duo, plugin, containerSelector, option_name):
+    select_by_name(dash_duo, containerSelector, option_name)
+    dash_duo.wait_for_element(f"#{plugin.uuid(option_name)}")
```
This indirection is not really required, I would remove it and use just `select_by_name`

valentin-krasontovitsch replied:
yass, you're right - done!

DanSava commented on tests/conftest.py:
```diff
@@ -100,3 +101,68 @@ def select_by_name(dash_duo, selector, name):
     return parameter_selector_input.text.split("\n")
+
+
+def setup_plugin(dash_duo, name, pluginClass, window_size=(1024, 3600)):
```
Seem to be very similar to `tests/data/spe1_st/tests/test_webviz_ert.py: _gen_plugin` maybe try to also replace `_gen_plugin` with this new function to keep code duplication minimal.

valentin-krasontovitsch replied:
thank you, i had seen it, and was later looking for it - but didn't immediately find it :sweat_smile: refactored as you suggested!

DanSava commented on webviz_ert/controllers/multi_response_controller.py:
```diff
@@ -196,4 +202,34 @@ def _generate_plot(ensemble_id: str, color: str) -> Optional[ResponsePlotModel]:
         return fig
+
+
+def get_index_type_with_sample(
```
The function returns multiple return variables, we should try to avoid that. Also not really sure what this function adds, it seems that what it returns is only used by the new function `axis_label_for_type` and also the logic is not really that complex so I guess it can just be part of `axis_label_for_type`.

DanSava commented on tests/data/snake_oil_data.py:
```diff
@@ -12,9 +12,20 @@ def content(self):
         return self._content
 
 
+ENSEMBLE_NAME_3 = "default3"
+INDEX_RESPONSE = "FGPT"
+DATE_RESPONSE = "FOPR"
```
I think if we have constants defined in a file they should be first after the import statements.

valentin-krasontovitsch replied:
thank you, fixed! surprising that neither black nor pylint pick up on that... apparently the python style guide doesn't have an opinion about it. huh

DanSava commented on tests/views/test_plot_view.py:
```diff
@@ -112,3 +118,42 @@ def test_clearing_ensembles_view(
+    )
+
+    # find the right ensemble which has mock data prepared for this test
+    wanted_ensemble_name = "default3"
```
I guess use the constant defined in `tests/data/snake_oil_data.py`

valentin-krasontovitsch replied:
thanks for catching that, fixed!

valentin-krasontovitsch commented on webviz_ert/controllers/multi_response_controller.py:
```diff
@@ -196,4 +202,34 @@ def _generate_plot(ensemble_id: str, color: str) -> Optional[ResponsePlotModel]:
+    if index_type == pd.core.indexes.numeric.Int64Index:
+        return "Index"
+
+    print(
+        f"got type `{index_type}` and data sample `{data_sample}` - not"
+        " sure what that is"
+    )
```
thank you, yes, of course - I'll use the logging module like in `webviz_ert/models/plot_model.py`

valentin-krasontovitsch commented on tests/data/snake_oil_data.py:
```diff
@@ -12,9 +12,20 @@ def content(self):
 all_ensemble_names = [
     "default",
-    "default3",
+    ENSEMBLE_NAME_3,
```
just created an [issue for the rest of the refactoring](https://github.com/equinor/webviz-ert/issues/278)!

valentin-krasontovitsch commented on tests/views/test_response_correlation.py:
```diff
@@ -48,3 +57,54 @@ def test_response_selector_sorting(mock_data, dash_duo):
     assert response_list[3] == "test_response_99"
+
+
+def test_axes_labels(mock_data, dash_duo):
```
they don't quite - I fixed `Comparison` first, and later (while reviewing my changes) noticed that `Correlation` was not fixed! Each Controller required fixing. I guess that should maybe be refactored...

valentin-krasontovitsch replied:
it could be refactored into a function that get's a loaded ensemble, a response (name), and a figure, and has the side effect of applying axes labels to the figure based on the index of the data of the given ensemble and response, and said function could be used in both controllers - but we'll leave it as is now.

valentin-krasontovitsch commented:
jenkins: test this please

hnformentin commented on tests/data/spe1_st/tests/test_webviz_ert.py:
```diff
@@ -6,23 +6,12 @@
-    dash_duo_handle.driver.set_window_size(*windowsize)
-    return plugin
-
-
```
The PR is marking as changes when you delete a function from one file and adds it to another one, e.g. `setup_plugin` and `_gen_plugin` (they are basically the same, just change the function name and file). This may add noise to the PR, as it seems large than it actually is. Do you know a way to make this kind of change cleaner? We can also discuss within the team to check the opportunities there!

oysteoh replied:
if you look at the two commits there is one for the refactoring and one for the actual change. I think that makes a nice separation at least

valentin-krasontovitsch replied:
yes, you're right of course, @hnformentin - i was perhaps a bit too eager in this PR, partly because I didn't take the time to read all of the surrounding code. we decided earlier that splitting it into two commits and limiting the refactoring should be acceptable.
i promise to refrain from refactoring in issue PRs in the future : )

hnformentin commented:
Regarding the labels, I am not sure too. We can discuss in the standup. But I think that indicating the project and squad is part of letting clean projects (after adding project, you need to wait a while to load the squads 🍰 ).

oysteoh commented on tests/data/snake_oil_data.py:
```diff
@@ -3,6 +3,11 @@
 import pandas as pd
 
 
+ENSEMBLE_NAME_3 = "default3"
```
really not sure if i like these name refactorings in the actual PR. For me they are messing up with raw test data as json, but i'm open for other opinions of course. To begin with this was dumping text from the endpoints, and i think it should be possible to to this in the future as well if needed... :man_shrugging:

oysteoh commented on webviz_ert/controllers/multi_response_controller.py:
```diff
@@ -196,4 +202,32 @@ def _generate_plot(ensemble_id: str, color: str) -> Optional[ResponsePlotModel]:
+    index_type = type(index)
+    data_sample = index[0]
+    if index_type == pd.core.indexes.base.Index:
+        # try to convert to date-time
```
this comment is not necessary as i think the code explains itself :+1:

valentin-krasontovitsch replied:
indeed : ) fixed!

oysteoh commented on webviz_ert/controllers/multi_response_controller.py:
```diff
@@ -196,4 +202,32 @@ def _generate_plot(ensemble_id: str, color: str) -> Optional[ResponsePlotModel]:
+        return "Index"
+
+    logger.warning(
+        f"got type `{index_type}` and data sample `{data_sample}` - not"
```
Maybe add some more information to this warning to help us figure out what is happening. I.e i would guess it would be beneficial having response name / ensemble name available.

valentin-krasontovitsch replied:
done and done.

should I add context as well, like that we're trying to figure out an axis label based on the index? prepend with the function name? or is that information given by the logger?
or it perhaps too much information? or is one supposed to find this info by just searching in the code, if one is interested?

oysteoh commented on webviz_ert/controllers/multi_response_controller.py:
```diff
@@ -196,4 +202,32 @@ def _generate_plot(ensemble_id: str, color: str) -> Optional[ResponsePlotModel]:
+
+
+def axis_label_for_ensemble_response(
+    ensemble: "EnsembleModel", responseName: str
```
should use snake_case: responseName -> response_name

valentin-krasontovitsch replied:
fixed, thank you!

oysteoh commented:
> Regarding the labels, I am not sure too. We can discuss in the standup. But I think that indicating the project and squad is part of letting clean projects (after adding project, you need to wait a while to load the squads cake ).

Labels should communicate what solving the issue would do. Milestone is for issues describing the actual milestone, i.e #243 . Adding axes labels is definitely an improvement / enhancement ( really not sure what is the different between those two - maybe a cleanup of the labels should be done.... )

valentin-krasontovitsch commented:
@DanSava ya i'm seeing that as well - gonna check it out!

valentin-krasontovitsch commented:
@DanSava the bug arises, as far as i understand, when there is no data for the response (more precisely the index is an empty list or whatever its structure is)

i have added a catch condition for when the index is empty - it does the trick in the case you point out, no error message, and an empty plot.

valentin-krasontovitsch commented:
jenkins, please test this

valentin-krasontovitsch commented:
Jenkins: Test this please!

valentin-krasontovitsch commented:
I am still at a loss as to why the komodo build by jenkins fails. and i'm not entirely sure how to approach that, as the logs don't seem to tell me much.

But besides that, I think that this might be ripe for merging now - I think I fixed the bug you mentioned, @DanSava [should I add a regression unit test, now that i think of it!?], and i've removed the refactoring that I did with string literals in the test data and test scripts.

valentin-krasontovitsch commented:
Jenkins: test this please.

DanSava commented on tests/data/spe1_st/tests/test_webviz_ert.py:
```diff
@@ -122,7 +111,7 @@ def test_webviz_response_comparison(dash_duo):
 @pytest.mark.spe1
 def test_webviz_observation_analyzer(dash_duo):
-    plugin = _gen_plugin(dash_duo, ObservationAnalyzer)
+    plugin = setup_plugin(dash_duo, ObservationAnalyzer)
```
same as above

DanSava commented on tests/data/spe1_st/tests/test_webviz_ert.py:
```diff
@@ -86,7 +75,7 @@ def test_webviz_response_correlation(dash_duo):
 @pytest.mark.spe1
 def test_webviz_response_comparison(dash_duo):
-    plugin = _gen_plugin(dash_duo, ResponseComparison)
+    plugin = setup_plugin(dash_duo, ResponseComparison)
```
same as above

valentin-krasontovitsch replied:
fixed

DanSava commented on tests/data/spe1_st/tests/test_webviz_ert.py:
```diff
@@ -54,7 +43,7 @@ def test_webviz_parameter_comparison(get_ensemble_id, dash_duo):
 @pytest.mark.spe1
 def test_webviz_response_correlation(dash_duo):
-    plugin = _gen_plugin(dash_duo, ResponseCorrelation)
+    plugin = setup_plugin(dash_duo, ResponseCorrelation)
```
same as above

valentin-krasontovitsch replied:
fixed

DanSava commented on tests/data/spe1_st/tests/test_webviz_ert.py:
```diff
@@ -37,7 +26,7 @@ def _verify_keys_in_menu(dash_duo_handle, plugin, keys, selector):
     # here we need the poke storage first - to get this test running
     _ = get_ensemble_id
-    plugin = _gen_plugin(dash_duo, ParameterComparison)
+    plugin = setup_plugin(dash_duo, ParameterComparison)
```
I think you are missing the `name` argument and also it might be a good idea to set the window size to what it was before the change `windowsize = (630, 2000)`

valentin-krasontovitsch replied:
ah thanks, of course! fixed. gonna run tests, and repush

valentin-krasontovitsch replied:
that is so funny - the tests seem to work out just fine, but jenkins actually catches the missing name argument O_o

valentin-krasontovitsch replied:
fixed!

DanSava reviewed (commented):
Looks much better :+1: but still needs a small fix in `test_webviz_ert.py`
